### PR TITLE
feat: split props and slots in vue 3

### PR DESCRIPTION
**Pull request by lee-chase**
Issue: #12850 

## What I did

Added non-enumerable sections (containing slots, props etc) to the results of extractArgTypes in Vue 3. 

In Vue 3 the current example extracts all args and binds them to the component. This means that not only props are extracted but also any other values such as slots. The current system also has an issue with slots and props with identical names.

Splitting them this way without affecting the current result usage, so is non-breaking.

## How to test

- All sections should appear as a non-enumerable property of args
- All slots should be in args.slots
- All props should be in args.props

- Does this need a new example in the kitchen sink apps?

Yes, we should recommend using args.slots and args.props.

- Does this need an update to the documentation?

Yes, we should recommend using args.slots and args.props.

If your answer is yes to any of these, please make sure to include it in your PR.

Apologies, it's Easter weekend. I Will try to catch up on docs/kitchen sink early next week but wanted to submit as alternative to #14465

### CODE REVIEW

File: addons/docs/src/frameworks/vue3/extractArgTypes.ts [hunk 1]
```diff
@@ -33,6 +33,13 @@ export const extractArgTypes: ArgTypesExtractor = (component) => {
           category: section,
         },
       };
+
+      // split props, slots etc which could clash on name
+      if (results.section === undefined) {
```
Comment (TimonLukas) [inline addons/docs/src/frameworks/vue3/extractArgTypes.ts:38]: This will be `undefined` every time, you have to check for the key: `results[section]`

**Review by TimonLukas — COMMENTED** [review]
I like this solution, a lot better than mine. Only thing I'd change: If the slot data etc. gets optionally exposed through a different API, I would not expect props to be overwritten. If we add a check for that around the `results[name]` part it would IMO lead to very intuitive behavior:

- If you only have a prop OR slot with a given name, everything works as it used to.
- If you have a prop and slot with the same name, the prop will beused, and the slot will only be exposed in the `slots` argument.

This feels more natural to me, and also is more useful regarding binding `args` to the component and using the `slots` from a different argument.

File: addons/docs/src/frameworks/vue3/extractArgTypes.ts (hunk 1 above)
Comment (lee-chase) [reply addons/docs/src/frameworks/vue3/extractArgTypes.ts]: Good catch, building storybook was a bit slow for me.

File: addons/docs/src/frameworks/vue3/extractArgTypes.ts [hunk 2]
```diff
@@ -33,6 +33,13 @@ export const extractArgTypes: ArgTypesExtractor = (component) => {
           category: section,
         },
       };
+
+      // split props, slots etc which could clash on name
+      if (results[section] === undefined) {
+        // add non-enumerable property for section
+        Object.defineProperty(results, section, { value: {}, enumerable: false });
+      }
+      results[section][name] = results[name];
```
Comment (shilman) [inline addons/docs/src/frameworks/vue3/extractArgTypes.ts:42]: What happens if there is a `prop` called `"slots"`?

**Review by shilman — COMMENTED** [review]
Hi @lee-chase! Thanks for putting this together! I was thinking about something similar as a general solution for all frameworks that have the opportunity for collisions.

WDYT about the following:
1. We make this a breaking change, and put it behind an opt-in feature flag like `scopedArgs: true`, which becomes the default in 7.0.
2. All non-props (slots, events, etc.) get scoped under `results[section]` as you have done here and removed from `results`, so we don't need to worry about collisions.

I can figure out how to get this documented applied to all other frameworks that need it.

We're in a stabilization period for 6.2, so won't be merging any 6.3 changes for at least a week.

File: addons/docs/src/frameworks/vue3/extractArgTypes.ts (hunk 2 above)
Comment (lee-chase) [reply addons/docs/src/frameworks/vue3/extractArgTypes.ts:42]: Good point and I guess depending on processing order a slot called 'props'.

Perhaps with the switch 'scopedArgs' the roll-up of props does not happen or goes into `results.all` as a fallback.

Not sure I like that as a solution, but I haven't thought of a better one just yet. 

Using non-permitted attribute name characters is another possibility, but I don't know enough about other frameworks to know where the clashes might be. `scopedArgs: true` could become `scopedArgs: { prefix: '$' }`

**Comment by braddialpad** [comment]
Hey guys any updates on this one? Happy to help out as well if there's any way I can.

File: addons/docs/src/frameworks/vue3/extractArgTypes.ts (hunk 2 above)
Comment (ndelangen) [reply addons/docs/src/frameworks/vue3/extractArgTypes.ts:42]: @fallemand are you saying we should not merge this PR as-is? Or that we should.

Sorry 😅 I don't quite understand.

Comment (fallemand) [reply addons/docs/src/frameworks/vue3/extractArgTypes.ts:42]: @ndelangen I'm saying we SHOULD merge it 🙏 .  
The issue is a nightmare. It's a common pattern in Vue design systems to use slots and props with the same name.

**Comment by shilman** [comment]
I've proposed what I believe to be a superior solution here https://github.com/storybookjs/storybook/pull/18407#discussion_r892143929

cc @ndelangen @prashantpalikhe

**Comment by ndelangen** [comment]
@lee-chase Is this PR made redundant by the merger of this PR: https://github.com/storybookjs/storybook/pull/20536 ?

**Pull request closed without merging**